### DISCUSSION TITLE
Problems with * character

### DISCUSSION THREAD

**Discussion by koimark** [dirkwhoffmann/virtualc64#743] [Q&A]
I have Finnish version of macOS - now newest Ventura but problems was earlier too - using VICE instead but VirtualC64 offers better UI so I occasionally try again. I have changed character roms and keyboard layouts etc but same happens: * character is missing I got @ instead. I even tried positional mapping and choose key manually but still in emulator I got @. That annoyes because menu option "Load First File" I got:
"Load "@",8,1" instead the correct one "Load"*",8,1
I have used character roms found in zimmers.net

**Comment by mithrendal** [comment]
hi @koimark could you please make a photo of your physical keyboard and paste that into here ? That would help to identify what is going on with the * key 🙃

**Comment by koimark** [comment]
Here is picture of keyboard. *-character is left side o return - top of '-character. But when try that in emulator I got the same @-character. (Normal place for @-characters is OPTION+2 - that does not do anything in emulator)
In VICE I got *-character  in keyboard just over the key marked as * 

![2022-12-07 22 48 53](https://user-images.githubusercontent.com/48367975/206301341-51b9abc9-caa7-4fe2-9a77-0b97e22acabc.jpg)

Keyboard information from system report:
Nykyinen syöttölähde:	com.apple.keylayout.Finnish
Kielikoodi:	fi
Sijainti:	fi_FI

**Comment by dirkwhoffmann** [comment]
Hmm, that's strange (and probably hard to debug without a Finnish keyboard). In symbolic mode, it shouldn't make any difference where the `*` key is located. In this mode, VirtualC64 processes characters, not key codes. What had to be done is to set a breakpoint in the following function of the `KeyboardController` class and step through it  in Xcode after `*` has been pressed:

```swift
   func keyDown(with event: NSEvent) {
                
        // Intercept if the console is open
        if renderer.console.isVisible { renderer.console.keyDown(with: event); return }

        // Ignore repeating keys
        if event.isARepeat { return }
        
        // Exit fullscreen mode if escape key is pressed
        if event.keyCode == kVK_Escape && renderer.fullscreen && pref.exitOnEsc {
            parent.window!.toggleFullScreen(nil)
            return
        }
        
        // Ignore keys that are pressed in combination with the Command key
        if event.modifierFlags.contains(NSEvent.ModifierFlags.command) { return }
        
        keyDown(with: MacKey(event: event))
    }
```

**Reply by koimark** [reply]
Thanks for the effort. I installed Xcode and download project but my skills run out to use breakpoint. I will try to learn more later.
But I tried to change my keyboard to GB and when i input characters "[ ] ; ’ \ , . / " in terminal I got results like in screenshot. Interesting is that in VirtualC64 input is ]-character when in macOS there is [-character  and ]-character comes to VirtualC64 when pressing next rows ;- key

<img width="196" alt="Näyttökuva 2022-12-9 kello 20 57 49" src="https://user-images.githubusercontent.com/48367975/206776086-46eee84d-4be6-47ec-b41f-bca99f4bb253.png">

**Reply by mithrendal** [reply]
Just out of curiosity. Do you get the same results when typing the mentioned keys in the web version of virtualC64? 

https://vc64web.github.io

**Reply by koimark** [reply]
Thanks for reply. That caused me to find a solution and the problem is solved. Problem was caused by kernal.325017.swedish-02.bin -rom I used as Kernal rom.
